File: code/video/MyCloudMusicAndroidJava/app/src/main/java/com/ixuea/courses/mymusic/component/download/fragment/DownloadingFragment.java
package com.ixuea.courses.mymusic.component.download.fragment;

import android.os.Bundle;

import com.ixuea.android.downloader.domain.DownloadInfo;
import com.ixuea.courses.mymusic.R;
import com.ixuea.courses.mymusic.component.download.adapter.DownloadingAdapter;
import com.ixuea.courses.mymusic.component.download.repository.DownloadRepository;
import com.ixuea.courses.mymusic.databinding.FragmentDownloadingBinding;
import com.ixuea.courses.mymusic.fragment.BaseViewModelFragment;
import com.ixuea.superui.toast.SuperToast;
import com.ixuea.superui.util.SuperRecyclerViewUtil;

import java.util.List;

/**
 * 下载中界面
 */
public class DownloadingFragment extends BaseViewModelFragment<FragmentDownloadingBinding> {
    private DownloadingAdapter adapter;
    private DownloadRepository repository;

    @Override
    protected void initViews() {
        super.initViews();
        SuperRecyclerViewUtil.initVerticalLinearRecyclerView(binding.list);
    }


    @Override
    protected void initDatum() {
        super.initDatum();
        repository = DownloadRepository.getInstance();

        //创建适配器
        adapter = new DownloadingAdapter(getHostActivity(), getOrm(), getChildFragmentManager());

        //设置适配器
        binding.list.setAdapter(adapter);

        loadData();
    }

    @Override
    protected void initListeners() {
        super.initListeners();
        adapter.setOnItemClickListener((holder, position) -> {
            //获取点击的数据
            DownloadInfo data = adapter.getData(position);

            switch (data.getStatus()) {
                case DownloadInfo.STATUS_NONE:
                case DownloadInfo.STATUS_PAUSED:
                case DownloadInfo.STATUS_ERROR:
                    //继续下载
                    repository.resume(data);
                    break;
                default:
                    //暂停下载
                    repository.pause(data);
                    break;
            }

            //显示按钮状态
            showButtonStatus();
        });

        binding.download.setOnClickListener(v -> downloadClick());
        binding.delete.setOnClickListener(v -> deleteClick());
    }

    private void deleteClick() {
        if (adapter.getItemCount() == 0) {
            SuperToast.show(R.string.error_not_download);
            return;
        }

        //删除所有下载任务
        for (DownloadInfo downloadInfo : adapter.getDatum()
        ) {
            repository.remove(downloadInfo);
        }

        //清除适配器数据
        adapter.clearData();
    }

    private void downloadClick() {
        if (adapter.getItemCount() == 0) {
            SuperToast.show(R.string.error_not_download);
            return;
        }

        if (isDownloading()) {
            pauseAll();
        } else {
            resumeAll();
        }

        //显示按钮状态
        showButtonStatus();
    }

    /**
     * 显示按钮状态
     */
    private void showButtonStatus() {
        if (isDownloading()) {
            binding.download.setText(R.string.pause_all);
        } else {
            binding.download.setText(R.string.download_all);
        }
    }


    private void resumeAll() {
        repository.resumeAll();
        adapter.notifyDataSetChanged();
    }

    private void pauseAll() {
        repository.pauseAll();
        adapter.notifyDataSetChanged();
    }

    /**
     * 是否有下载任务
     *
     * @return
     */
    private boolean isDownloading() {
        return repository.isDownloading(adapter.getDatum());
    }

    @Override
    protected void loadData(boolean isPlaceholder) {
        super.loadData(isPlaceholder);
        List<DownloadInfo> downloads = repository.findDownloading();
        adapter.setDatum(downloads);

        //显示按钮状态
        showButtonStatus();
    }

    public static DownloadingFragment newInstance() {

        Bundle args = new Bundle();

        DownloadingFragment fragment = new DownloadingFragment();
        fragment.setArguments(args);
        return fragment;
    }
}

File: code/video/MyCloudMusicAndroidJava/app/src/main/java/com/ixuea/courses/mymusic/component/download/fragment/DownloadingFragment.kt
package com.ixuea.courses.mymusic.component.download.fragment

import android.os.Bundle
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.ixuea.android.downloader.domain.DownloadInfo
import com.ixuea.courses.mymusic.R
import com.ixuea.courses.mymusic.component.download.adapter.DownloadingAdapter
import com.ixuea.courses.mymusic.component.download.ui.DownloadingUiState
import com.ixuea.courses.mymusic.component.download.ui.DownloadingViewModel
import com.ixuea.courses.mymusic.databinding.FragmentDownloadingBinding
import com.ixuea.courses.mymusic.fragment.BaseViewModelFragment
import com.ixuea.superui.toast.SuperToast
import com.ixuea.superui.util.SuperRecyclerViewUtil
import kotlinx.coroutines.launch

/**
 * 下载中界面
 */
class DownloadingFragment : BaseViewModelFragment<FragmentDownloadingBinding>() {
    private lateinit var adapter: DownloadingAdapter
    private lateinit var viewModel: DownloadingViewModel
    private var handledDataVersion = 0L

    override fun initViews() {
        super.initViews()
        SuperRecyclerViewUtil.initVerticalLinearRecyclerView(binding.list)
    }

    override fun initDatum() {
        super.initDatum()
        viewModel = ViewModelProvider(this)[DownloadingViewModel::class.java]

        adapter = DownloadingAdapter(hostActivity, orm, childFragmentManager)
        adapter.setListener(object : DownloadingAdapter.DownloadingAdapterListener {
            override fun onDeleteClick(position: Int, data: DownloadInfo) {
                viewModel.remove(data)
            }

            override fun onDownloadTerminalState(data: DownloadInfo) {
                viewModel.onDownloadTerminalState()
            }
        })
        binding.list.adapter = adapter
        observeDownloadingState()

        loadData()
    }

    override fun initListeners() {
        super.initListeners()
        adapter.setOnItemClickListener { _, position ->
            val data = adapter.getData(position)
            viewModel.toggle(data)
        }

        binding.download.setOnClickListener { downloadClick() }
        binding.delete.setOnClickListener { deleteClick() }
    }

    private fun observeDownloadingState() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state ->
                    render(state)
                }
            }
        }
    }

    private fun render(state: DownloadingUiState) {
        if (state.dataVersion != handledDataVersion) {
            handledDataVersion = state.dataVersion
            adapter.setDatum(state.downloads)
            showButtonStatus(state.isDownloading)
        }
    }

    private fun deleteClick() {
        if (adapter.itemCount == 0) {
            SuperToast.show(R.string.error_not_download)
            return
        }

        viewModel.removeAll(adapter.datum.toList())
    }

    private fun downloadClick() {
        if (adapter.itemCount == 0) {
            SuperToast.show(R.string.error_not_download)
            return
        }

        if (viewModel.uiState.value.isDownloading) {
            viewModel.pauseAll()
        } else {
            viewModel.resumeAll()
        }
    }

    private fun showButtonStatus(isDownloading: Boolean) {
        if (isDownloading) {
            binding.download.setText(R.string.pause_all)
        } else {
            binding.download.setText(R.string.download_all)
        }
    }

    override fun loadData(isPlaceholder: Boolean) {
        super.loadData(isPlaceholder)
        viewModel.load()
    }

    companion object {
        @JvmStatic
        fun newInstance(): DownloadingFragment {
            return DownloadingFragment().apply {
                arguments = Bundle()
            }
        }
    }
}
